specification/resources/resource-manager/Microsoft.Resources/preview/2024-06-01-preview/changes.json

```diff
@@ -0,0 +1,408 @@
+          "type": "string",
+          "readOnly": true
+        },
+        "isTruncated": {
+          "description": "The flag indicating whether property changes dictionary was truncated in this resource",
+          "type": "boolean",
+          "readOnly": true
+        },
+        "changedBy": {
+          "description": "The identifier i.e. email id or application id of the entity that made the change(s)",
+          "type": "string",
+          "readOnly": true
+        },
+        "changedByType": {
+          "description": "The entity type that made the change(s)",
+          "type": "string",
+          "readOnly": true,
+          "enum": [
+            "Application",
+            "User",
+            "ManagedIdentity",
+            "Unknown"
+          ],
+          "x-ms-enum": {
+            "name": "changedByType",
+            "modelAsString": true,
+            "values": [
+              {
+                "value": "Application",
+                "description": "An application made the change(s)",
+                "name": "Application"
+              },
+              {
+                "value": "User",
+                "description": "A user made the change(s)",
+                "name": "User"
+              },
+              {
+                "value": "ManagedIdentity",
+                "description": "A ManagedIdentity made the change(s)",
+                "name": "ManagedIdentity"
+              },
+              {
+                "value": "Unknown",
+                "description": "An Unknown entity made the change(s)",
+                "name": "Unknown"
+              }
+            ]
+          }
+        },
+        "clientType": {
+          "description": "Client used to make the changes",
+          "type": "string",
+          "readOnly": true
+        },
+        "operation": {
+          "description": "The Azure Resource Provider operation that was invoked to make the change(s). See [here](https://learn.microsoft.com/azure/role-based-access-control/resource-provider-operations) for a full list of possible values",
+          "type": "string",
+          "readOnly": true
+        },
+        "changeControlId": {
+          "description": "The resource identifier of the change control resource associated with this change record.",
+          "type": "string",
+          "format": "arm-id",
+          "readOnly": true
+        }
+      }
+    },
+    "ChangesDictionary": {
+      "description": "A dictionary with changed property name as a key and the change details as the value",
+      "type": "object",
+      "additionalProperties": {
+        "$ref": "#/definitions/ChangeBase"
+      }
+    },
+    "ChangeBase": {
+      "description": "An individual change on the target resource",
+      "type": "object",
+      "properties": {
+        "previousValue": {
+          "description": "The target resource property value before the change",
+          "type": "string",
```

Comment:
What if the value before wasn't a string, but an integer? How do you distinguish between integer value of 4, and string value of "4"? Null and "null"? Or just not defined, and null? (which might be represented differently for some resources!) How do you represent array changes?

Reply:
it is an optional/nullable string type, other than null all values including integers/json/xml etc.  are converted to string

Reply:
Wouldn't that be quite inconvenient though, to consume programmatically, if you were trying to consume these 'changes' and do something like e.g. automatically replicate them on another resource, or roll them back?

You'd have to get the string AND also know the underlying type of the property you're trying to update.

Reply:
Yes it will need an extra step to parse the strings first for now, as it has been since quite a while.
We will consider your feedback in future iterations as it will be a breaking change. For now, this isn't       a blocking comment, so can we go ahead with merge (based on previous review, spanning now over a month)

Reply:
If its intentionally nullable, I believe you need to annotate it as "x-nullable": true to indicate that null is a legal value.

This is a breaking change.